sqlmesh/core/engine_adapter/base.py

```diff
@@ -67,6 +70,34 @@ def is_insert_overwrite(self) -> bool:
     def is_replace_where(self) -> bool:
         return self == InsertOverwriteStrategy.REPLACE_WHERE
 
+    @property
+    def requires_condition(self) -> bool:
+        return self.is_replace_where or self.is_delete_insert
+
+
+class SourceQuery:
+    def __init__(
+        self,
+        query_factory: t.Callable[[], Query],
+        cleanup_func: t.Optional[t.Callable[[], None]] = None,
+        **kwargs: t.Any,
+    ) -> None:
+        self.query_factory = query_factory
+        self.cleanup_func = cleanup_func
+
+    def __enter__(self) -> Query:
+        return self.query_factory()
+
+    def __exit__(
+        self,
+        exc_type: t.Optional[t.Type[BaseException]],
+        exc_val: t.Optional[BaseException],
+        exc_tb: t.Optional[types.TracebackType],
+    ) -> bool:
+        if self.cleanup_func:
+            self.cleanup_func()
+        return True
```

Comment:
> If an exception is supplied, and the method wishes to suppress the exception (i.e., prevent it from being propagated), it should return a true value. Otherwise, the exception will be processed normally upon exit from this method.

I don't think we should be suppressing an exception here.

Reply:
https://docs.python.org/3/reference/datamodel.html#object.__exit__